design/one-pager-class-selection.md

```diff
@@ -0,0 +1,371 @@
+1. If no `CloudSQLInstanceClass` matched the labels, the reconcile is done.
+   Otherwise, one of the matching `CloudSQLInstanceClass` resources is selected
+   at random.
+1. The scheduler sets the `classRef` of the `PostgreSQLInstance` to the selected
+   `CloudSQLInstanceClass`. If two controllers try to set the `classRef` at the
+   same time one will fail to commit the change due to the `PostgreSQLInstance`
+   claim's resource version having changed since it was read.
+1. The reconcile is done. With the `classRef` set the `PostgresSQLInstance` now
+   passes the watch predicates of the `(PostgreSQLInstance, CloudSQLInstance)`
+   resource claim reconciler, which dynamically provisions and binds it.
+
+All portable resource class definitions and defaulting controllers would be
+removed from Crossplane core.
+
+## Future Considerations
+
+It may be possible to improve upon the random scheduling of resource claims to
+resource classes by supporting weighted random scheduling. Under such a proposal
+resource classes would have a spec field or annotation specifying their weight,
+with the highest weight winning. This could be used by infrastructure
+administrators to ensure conservative (or affordable) behaviour in cases where a
+resource claim matches multiple resource classes.
+
+Consider the following resource claim:
+
+```yaml
+apiVersion: cache.crossplane.io/v1alpha1
+kind: RedisCluster
+metadata:
+  namespace: acme-team
+  name: gitlab-cache
+spec:
+  classSelector:
+    matchLabels:
+      stack: gitlab
+      region: us-west
+```
+
+Matched against the below resource classes:
+
+```yaml
+apiVersion: cache.gcp.crossplane.io/v1alpha2
+kind: CloudMemorystoreInstanceClass
+metadata:
+  name: really-quite-a-lot-of-memory
+  labels:
+    stack: gitlab
+    region: us-west
+    grade: production
+specTemplate:
+  tier: STANDARD_HA
+  region: us-west2
+  memorySizeGb: 99999999999999999999999
+  redisVersion: REDIS_3_2
+---
+apiVersion: cache.gcp.crossplane.io/v1alpha2
+kind: CloudMemorystoreInstanceClass
+metadata:
+  name: cheap-redis
+  labels:
+    stack: gitlab
+    region: us-west
+    grade: experimental
+specTemplate:
+  tier: STANDARD_HA
+  region: us-west2
+  memorySizeGb: 1
+  redisVersion: REDIS_3_2
+```
+
+Both classes match the resource claim's desired properties, but one is way more
+expensive than the other. Given the resource claim did not specifically request
+a production database, it should probably be scheduled to use the cheaper class.
+Weighting could be used to approximate the "cost" of resource classes - higher
+weights are cheaper (or safer) resource classes.
+
+Note that anything other than random scheduling would require the various class
+scheduler controllers to collaborate. In the case of weighted scheduling each
+controller would need to determine the highest weighted matching resource class,
+then determine whether its highest weight was higher than that chosen by any
+other scheduler controller.
+
```

Comment:
Another slightly pie-in-the-sky future consideration. If resource claims are matched to resource classes using labels, and don't expose any actual configuration fields for the resource claim author to set, is there a point in resource claims being of distinct kinds? i.e. Could we have simply `ResourceClaim` rather than `PostgreSQLInstance`?

This would look something like:

```yaml
apiVersion: crossplane.io/v1alpha1
kind: ResourceClaim
metadata:
  namespace: acme-team
  name: gitlab-cache
spec:
  classSelector:
    matchLabels:
      protocol: postgresql
      stack: gitlab
      region: us-west
```

This direction would remove the need to add new resource claim kinds to Crossplane core in order to support new kinds of managed resources, and would remove any distinction between "portable" and "non-portable" resource claims - you could dynamically provision a hypothetical `DynamoDB` using `ResourceClaim` just as easily is you could provision a `PostgreSQLInstance`.

Reply:
The generic notion of this idea is appealing, but I feel like this is just moving strongly typed fields and assumptions from the type implementation  to, `classSelector` as "weak" fields, which brings out analogous challenges we faced in generic resource classes.

Reply:
@negz one drawback to this is that stack developers basically have to coordinate with each other to say what the protocol value is that they are going standardize on. Since there is no centralized `PostrgreSQLInstance` in core crossplane, developers on stack-aws and stack-gcp need to make sure they both accept `protocol: postgresql` for `RDSInstance` and `CloudSQLInstance` respectively, and not one accepts `protocol: postgres` or something. I do like the idea in theory that we do not have to add new claim types though (plus less CRDs is always a plus IMO). There would be incentive for stack developers to standardize with this pattern, but I am not sure we want to put the burden on them to look around and see what everyone else is doing rather than just knowing what core crossplane claim types are available.

Reply:
I think I'm going to close this thread out to prevent it distracting too much from the rest of the proposal. Definitely agree with the feedback.